File: app/src/main/java/com/example/earthwallet/chain/Personhood.kt
package network.erth.wallet.chain

import com.google.protobuf.ByteString
import network.erth.earth.proto.personhood.MsgClaimAnml
import network.erth.earth.proto.personhood.MsgRegister
import network.erth.wallet.Constants
import network.erth.wallet.wallet.services.EarthWallet
import org.bitcoinj.core.ECKey
import org.json.JSONObject

/**
 * x/personhood — proof-of-personhood registration and the daily ANML claim.
 *
 * The one-human-one-vote allocation stream this module gates lives in
 * x/allocation (see [Allocation], STREAM_ID_HUMAN). This module only decides who
 * counts as a live human; the votes and options belong to the allocation module.
 */
object Personhood {

    fun isRegistered(address: String): Boolean {
        val (code, body) = EarthRest.get("/earth-network/earth/personhood/v1/registration/$address")
        if (code !in 200..299) return false
        return JSONObject(body).optBoolean("registered", false)
    }

    /** Registration status, including the last ANML-claim time (unix seconds, 0 if never). */
    data class RegistrationStatus(val registered: Boolean, val expired: Boolean, val lastAnmlClaim: Long)

    fun registrationStatus(address: String): RegistrationStatus {
        val (code, body) = EarthRest.get("/earth-network/earth/personhood/v1/registration/$address")
        if (code !in 200..299) return RegistrationStatus(false, false, 0L)
        val json = JSONObject(body)
        val reg = json.optJSONObject("registration")
        return RegistrationStatus(
            registered = json.optBoolean("registered", false),
            expired = json.optBoolean("expired", false),
            lastAnmlClaim = reg?.optString("last_anml_claim", "0")?.toLongOrNull() ?: 0L,
        )
    }

    /** ANML is claimable at most once per UTC day (chain rule: now - last_anml_claim >= 86400). */
    fun isAnmlClaimable(status: RegistrationStatus): Boolean {
        if (!status.registered) return false
        val nowSec = System.currentTimeMillis() / 1000
        return nowSec - status.lastAnmlClaim >= 86400
    }

    /**
     * How many humans are currently registered — the denominator of the human
     * emission stream, since every registration carries the same weight.
     *
     * This used to ride along on the democratic-options response; it is its own
     * query now that the options belong to x/allocation.
     */
    fun registrationCount(): Long {
        val (code, body) = EarthRest.get("/earth-network/earth/personhood/v1/registration_count")
        if (code !in 200..299) return 0L
        return JSONObject(body).optString("count", "0").toLongOrNull() ?: 0L
    }

    /** One issuing country's registration total. */
    data class CountryCount(val country: String, val count: Long)

    /**
     * Registrations per issuing country, largest first. `country` is an ISO
     * 3166-1 alpha-2 code, or "" when the Document Signer's certificate carries
     * no country attribute. Powers the explorer's registrations tab.
     */
    fun registrationCountries(): List<CountryCount> {
        val (code, body) = EarthRest.get("/earth-network/earth/personhood/v1/registration_countries")
        if (code !in 200..299) return emptyList()
        val arr = JSONObject(body).optJSONArray("countries") ?: return emptyList()
        val out = ArrayList<CountryCount>(arr.length())
        for (i in 0 until arr.length()) {
            val c = arr.getJSONObject(i)
            out.add(
                CountryCount(
                    country = c.optString("country", ""),
                    count = c.optString("count", "0").toLongOrNull() ?: 0L,
                )
            )
        }
        return out.sortedByDescending { it.count }
    }

    /** How many humans registered with a given Document Signer (hex dsc_key). */
    fun registrationsByDsc(dscKeyHex: String): Long {
        val key = dscKeyHex.removePrefix("0x")
        val (code, body) = EarthRest.get(
            "/earth-network/earth/personhood/v1/registrations_by_dsc/$key"
        )
        if (code !in 200..299) return 0L
        return JSONObject(body).optString("count", "0").toLongOrNull() ?: 0L
    }

    // --- messages ---

    /**
     * Client-side proof-of-personhood registration. Returns tx hash.
     *
     * proof is the Barretenberg UltraHonk proof bytes; publicSignals are the
     * circuit public signals as decimal strings ([current_date, nullifier,
     * dsc_key] for lean_poa); signatureAlgorithm selects the on-chain verifying
     * key; dscDer is the Document Signer certificate the chain verifies against
     * its CSCA trust store and binds to the proof's dsc_key output.
     */
    /**
     * Registration's gas limit and fee, defined here — beside the message they
     * pay for — because the caller cannot set them by being careful.
     *
     * This existed as a constant in PassportScannerFragment that fed only the
     * confirmation sheet, while register() fell through to EarthTx.broadcast's
     * 400_000 default. Raising the fragment's copy changed the number the user
     * was shown and not the number sent, and the transaction still ran out of
     * gas at exactly the old limit.
     *
     * MsgRegister verifies an UltraHonk proof on-chain and is the most expensive
     * message the app sends. A fresh account pays more than a used one: the ante
     * handler stores its public key on the first transaction, which measured as
     * 400324 gas against a 400000 limit — over, and precisely the case that
     * matters, since a new human's first transaction is always this one.
     *
     * Generous rather than tuned: the fee is flat rather than gas x price, and
     * the chain reports max_gas -1, so headroom costs nothing while an
     * under-estimate burns the fee and the ad view that paid for it.
     */
    const val REGISTER_GAS_LIMIT = 3_000_000L
    const val REGISTER_FEE_UERTH = "2000"

    fun register(
        key: ECKey,
        proof: ByteArray,
        publicSignals: List<String>,
        signatureAlgorithm: String,
        affiliate: String?,
        dscDer: ByteArray,
    ): String {
        val msg = MsgRegister.newBuilder()
            .setCreator(EarthWallet.address(key))
            .setProof(ByteString.copyFrom(proof))
            .addAllPublicSignals(publicSignals)
            .setSignatureAlgorithm(signatureAlgorithm)
            .setAffiliate(affiliate ?: "")
            .setDscDer(ByteString.copyFrom(dscDer))
            .build()
        return EarthTx.broadcast(
            key,
            listOf(EarthTx.anyOf(Constants.MSG_REGISTER_TYPE_URL, msg)),
            gasLimit = REGISTER_GAS_LIMIT,
            feeUerth = REGISTER_FEE_UERTH,
        )
    }

    /** Daily ANML claim. Returns tx hash. */
    fun claimAnml(key: ECKey): String {
        val msg = MsgClaimAnml.newBuilder().setCreator(EarthWallet.address(key)).build()
        return EarthTx.broadcast(key, listOf(EarthTx.anyOf("/earth.personhood.v1.MsgClaimAnml", msg)))
    }
}

File: ios/EarthCore/Sources/EarthCore/Chain/Personhood.swift
import Foundation

/// x/personhood — registration and the daily ANML claim.
///
/// The one-human-one-vote stream this gates lives in x/allocation
/// (`StreamID.caretaker`). This module only decides who counts as a live human.
public enum Personhood {

    private static let secondsPerDay: Int64 = 86_400

    public struct RegistrationStatus: Sendable, Equatable {
        public let registered: Bool
        public let expired: Bool
        /// Unix seconds; 0 if never claimed.
        public let lastAnmlClaim: Int64

        public static let none = RegistrationStatus(registered: false, expired: false, lastAnmlClaim: 0)
    }

    /// ANML is claimable once per UTC day.
    ///
    /// Compared as day numbers, matching the chain. A rolling
    /// `now - last >= 86400` agrees only while the chain stores a
    /// midnight-truncated timestamp; against a real claim time it under-reports
    /// by up to a day, hiding a claim that is actually available.
    public static func isAnmlClaimable(_ status: RegistrationStatus, now: Date = Date()) -> Bool {
        guard status.registered else { return false }
        let nowSec = Int64(now.timeIntervalSince1970)
        return nowSec / secondsPerDay != status.lastAnmlClaim / secondsPerDay
    }

    /// Unix seconds at the next UTC midnight, when the window reopens.
    public static func nextClaimOpensAt(now: Date = Date()) -> Int64 {
        (Int64(now.timeIntervalSince1970) / secondsPerDay + 1) * secondsPerDay
    }

    /// One issuing country's registration total. `country` is ISO 3166-1
    /// alpha-2, or "" when the Document Signer's certificate carries no country.
    public struct CountryCount: Sendable, Equatable {
        public let country: String
        public let count: Int64
    }

    /// Registration's gas limit and fee, defined beside the message they pay for.
    ///
    /// MsgRegister verifies an UltraHonk proof on-chain and is the most
    /// expensive message the app sends. A fresh account pays more than a used
    /// one — the ante handler stores its public key on the first transaction,
    /// measured at 400324 gas against Android's old 400000 default. That is
    /// precisely the case that matters, since a new human's first transaction
    /// is always this one.
    ///
    /// Generous rather than tuned. Note that headroom is NOT free: the fee is
    /// `ceil(gas x minimum-gas-prices)`, so this limit costs 15,000 uerth at
    /// 0.005uerth whether or not the gas is used. The flat "2000" that used to
    /// sit here was the fee for 400,000 gas, and every registration was
    /// rejected with "insufficient fees; got: 2000uerth required: 15000uerth".
    /// An under-estimate still burns both the fee and the ad view that paid for
    /// it, so the headroom stays — but it is a priced decision, not a free one.
    public static let registerGasLimit: UInt64 = 3_000_000
    public static var registerFeeUerth: String { Fees.forGas(registerGasLimit) }
}

public extension EarthClient {

    func registrationStatus(_ address: String) async -> Personhood.RegistrationStatus {
        guard let json = try? await rest.get(
            "/earth/personhood/v1/registration/\(address)"
        ) else { return .none }
        return Personhood.RegistrationStatus(
            registered: json.registered.bool ?? false,
            expired: json.expired.bool ?? false,
            lastAnmlClaim: json.registration.last_anml_claim.int64(default: 0)
        )
    }

    func isRegistered(_ address: String) async -> Bool {
        await registrationStatus(address).registered
    }

    /// How many humans are registered — the denominator of the human emission
    /// stream, since every registration carries the same weight.
    func registrationCount() async -> Int64 {
        guard let json = try? await rest.get(
            "/earth/personhood/v1/registration_count"
        ) else { return 0 }
        return json.count.int64(default: 0)
    }

    /// Registrations per issuing country, largest first.
    func registrationCountries() async -> [Personhood.CountryCount] {
        guard let json = try? await rest.get(
            "/earth/personhood/v1/registration_countries"
        ) else { return [] }
        return json.countries.array
            .map {
                Personhood.CountryCount(
                    country: $0.country.string(default: ""),
                    count: $0.count.int64(default: 0)
                )
            }
            .sorted { $0.count > $1.count }
    }

    /// How many humans registered with a given Document Signer (hex dsc_key).
    func registrations(byDSC dscKeyHex: String) async -> Int64 {
        let key = dscKeyHex.hasPrefix("0x") ? String(dscKeyHex.dropFirst(2)) : dscKeyHex
        guard let json = try? await rest.get(
            "/earth/personhood/v1/registrations_by_dsc/\(key)"
        ) else { return 0 }
        return json.count.int64(default: 0)
    }

    // --- messages ---

    /// Proof-of-personhood registration.
    ///
    /// `proof` is the Barretenberg UltraHonk proof bytes; `publicSignals` are
    /// the circuit's public signals as decimal strings — `[current_date,
    /// nullifier, dsc_key]` for `lean_poa`; `signatureAlgorithm` selects the
    /// on-chain verifying key; `dscDer` is the Document Signer certificate the
    /// chain checks against its CSCA trust store and binds to the proof's
    /// `dsc_key` output.
    func msgRegister(
        creator: String,
        proof: Data,
        publicSignals: [String],
        signatureAlgorithm: String,
        affiliate: String?,
        dscDER: Data
    ) -> ProtoAny {
        Msg.Register(
            creator: creator,
            proof: proof,
            publicSignals: publicSignals,
            affiliate: affiliate ?? "",
            signatureAlgorithm: signatureAlgorithm,
            dscDer: dscDER
        ).asAny(typeURL: Msg.Register.typeURL)
    }

    @discardableResult
    func register(
        key: EarthKey,
        proof: Data,
        publicSignals: [String],
        signatureAlgorithm: String,
        affiliate: String?,
        dscDER: Data
    ) async throws -> String {
        try await broadcast(
            [msgRegister(
                creator: key.address,
                proof: proof,
                publicSignals: publicSignals,
                signatureAlgorithm: signatureAlgorithm,
                affiliate: affiliate,
                dscDER: dscDER
            )],
            key: key,
            gasLimit: Personhood.registerGasLimit,
            feeUerth: Personhood.registerFeeUerth
        )
    }

    func msgClaimAnml(creator: String) -> ProtoAny {
        Msg.ClaimAnml(creator: creator).asAny(typeURL: Msg.ClaimAnml.typeURL)
    }

    @discardableResult
    func claimAnml(key: EarthKey) async throws -> String {
        try await broadcast([msgClaimAnml(creator: key.address)], key: key)
    }

    // No unregister. The chain removed MsgUnregister — retiring a registration
    // freed its nullifier, and Register pays the registration reward to any
    // nullifier that is not already live, so leaving and returning drew on the
    // reward pool once per block. A registration now ends only by expiring, and
    // moves between wallets by registering again from the new one.
}
